### PR TITLE
Implement Admin delegation for group folder

### PR DESCRIPTION
**Pull request by CarlSchwan**
This is an alternative implementation of https://github.com/nextcloud/groupfolders/pull/1418
but using the new API available in NC 23 for admin delegation.

### CODE REVIEW

File: lib/Controller/DelegationController.php [hunk 1]
```diff
@@ -0,0 +1,67 @@
+<?php
+
+/**
+ * @author Cyrille Bollu <cyr.debian@bollu.be> for Arawa (https://www.arawa.fr/)
+ *
```
Comment (fschrempf) [inline lib/Controller/DelegationController.php:5]: Other files have a `@license` tag. Maybe we should also add it here.

Comment (CarlSchwan) [reply lib/Controller/DelegationController.php:5]: Good catch!

File: lib/Folder/FolderManager.php [hunk 2]
```diff
@@ -286,11 +286,21 @@ private function getGroups(int $id): array {
...
 	 * @throws Exception
 	 */
-	public function canManageACL(int $folderId, string $userId): bool {
+	public function canManageACL(int $folderId, IUser $user): bool {
+		$userId = $user->getUId();
 		if ($this->groupManager->isAdmin($userId)) {
 			return true;
 		}
 
+		// Call private server api
```
Comment (CarlSchwan) [inline lib/Folder/FolderManager.php:295]: Check for what this is needed and add comment

Comment (CarlSchwan) [reply lib/Folder/FolderManager.php:295]: Basically if this returns false the user can't manage the ACL in the admin UI or in the folder sidebar. Added a bit of doc

**Comment by fschrempf** [comment]
I gave this a quick local test and it seems to work just as expected!

**Comment by juliushaertl** [comment]
@StCyr Mind to see if this covers all use cases you had in mind with your previous PR, as you had a separate app integrating with your pull request?

I think going with the native delegation API that is now present with 23 is the best approach also to avoid larger chunks of additional code to maintain.

**Comment by zak39** [comment]
> @StCyr Mind to see if this covers all use cases you had in mind with your previous PR, as you had a separate app integrating with your pull request?
> 
> I think going with the native delegation API that is now present with 23 is the best approach also to avoid larger chunks of additional code to maintain.

Hi @juliushaertl :)

Yes, we must test this PR with Nextcloud 23 and check if it's okay for our app ^^

**Comment by CarlSchwan** [comment]
23 alpha should have my patch with the new IDelegatedSettings. I will try to look at that could cause this tomorrow

**Comment by zak39** [comment]
> 23 alpha should have my patch with the new IDelegatedSettings. I will try to look at that could cause this tomorrow

Thank you @CarlSchwan :)

Did you create the `IDelegatedSettings` interface ? :slightly_smiling_face: 

Can you let me know when you have information and when you can, please ? :slightly_smiling_face:

**Comment by CarlSchwan** [comment]
> > 23 alpha should have my patch with the new IDelegatedSettings. I will try to look at that could cause this tomorrow
> 
> Thank you @CarlSchwan :)
> 
> Did you create the `IDelegatedSettings` interface ? slightly_smiling_face

Yes, I created the IDelegatedSettings approximately two months ago. For me, this MR works perfectly and the class is here: https://github.com/nextcloud/server/blob/master/lib/public/Settings/IDelegatedSettings.php Could you make sure that it is also here on your local setup?

> Can you let me know when you have information and when you can, please ? slightly_smiling_face

**Comment by CarlSchwan** [comment]
Merging this since I have one approval and another user could confirm it worked fine on their system. I don't like having approved PRs sitting around. @zak39 once you are able to test, feel free to create issues if you have improvement ideas

**Comment by gdelafond** [comment]
Hello, 

Very nice !

The groupfolder version 11.0.0 that is available for NC23 (https://apps.nextcloud.com/apps/groupfolders) does not contains this PR, is there an ETA to push a new groupfolder version with this PR ?

Thank you.

**Comment by CarlSchwan** [comment]
IIRC @juliushaertl wanted to do an 11.1.0 release sometime soon, so stay tuned :)

**Pull request merged**